env/src/main/scala/tofu/env/bio/EnvBio.scala

```diff
@@ -0,0 +1,42 @@
+package tofu.env.bio
+
+import monix.eval.Task
+
+sealed trait EnvBio[-R, +E, +A] {
+  def run(ctx: R): Task[Either[E, A]]
+
+  def map[B](f: A => B): EnvBio[R, E, B] =
+    flatMap(value => EnvBio.pure(f(value)))
+
+  def flatMap[R1 <: R, E1 >: E, B](f: A => EnvBio[R1, E1, B]): EnvBio[R1, E1, B] =
+    EnvBio(ctx => run(ctx).flatMap(_.fold(raiseUserError, f(_).run(ctx))))
+
+  def mapTask[E1 >: E, B](f: Task[Either[E, A]] => Task[Either[E1, B]]): EnvBio[R, E1, B] =
+    EnvBio(ctx => f(run(ctx)))
+
+  private def raiseUserError[E1](e: E1): Task[Either[E1, Nothing]] = Task.raiseError(UserError(e))
+}
+
+object EnvBio extends EnvBioFunctions {}
+
+/** Context aware variation of `EnvBio` that allows specifying error type */
+final case class EnvBioCtx[R, +E, +A](runF: R => Task[Either[E, A]]) extends EnvBio[R, E, A] {
+  def run(ctx: R): Task[Either[E, A]] = Task.defer(runF(ctx)).attempt.flatMap {
+    case Left(UserError(err: E)) => Task.pure(Left(err))
```

Comment:
With this is encoding is this even possible that `runF` will ever raise `UserError` ?

Reply:
My initial idea was that computations can be short-circuited by `EnvBio.raiseError` that raises `UserError`. But now, I realise that `run` is called not only at the end of program, but can be called somewhere in the middle of computation, so that wasn't the best idea and it even prevents us from mapping on error channel. I will rework this to just passing Either around.

Reply:
I dont think people will like wrapping Eithers on every step

Reply:
I agree. Then internal encoding can be represented as `A => Task[B]` and user custom errors as `UserError <: Throwable`, so in error handling we will match error and handle user ones. Thanks!